src/oscar/navigation_menu_registry.py

```diff
@@ -0,0 +1,237 @@
+                break
+        else:
+            menu.auto_position_if_applicable(len(self._children))
+            self._children.append(menu)
+        return self
+
+    def add_children(self, children):
+        for child in children:
+            self.add_child(child)
+        return self
+
+    def remove_child(self, identifier):
+        try:
+            self._children.remove(self.placeholder(identifier))
+        except ValueError:
+            raise ValueError(f"Child menu with ID '{identifier}' not found for parent with ID '{self.identifier}'")
+        return self
+
+    def as_navigation_dict(self):
+        return {
+            "label": self.label,
+            "url_name": self.url_name,
+            "icon": self.icon,
+            "children": [menu.as_navigation_dict() for menu in self.children],
+            **self._use_as_provided,
+        }
+
+
+_dashboard_navigation_menu = defaultdict(Menu)
+_marked_for_deregistration = set()
+
+
+def get_legacy_parent_menu(lookup_label_or_index):
+    """Should help ease migration process from menus declared through `settings.OSCAR_DASHBOARD_NAVIGATION`"""
+    is_lookup_by_position = isinstance(lookup_label_or_index, int)
+    if is_lookup_by_position:
+        return settings.OSCAR_DASHBOARD_NAVIGATION[lookup_label_or_index]
+
+    # parent menu is to be identified by label
+    for index, parent_menu in enumerate(settings.OSCAR_DASHBOARD_NAVIGATION):
+        if parent_menu["label"] == lookup_label_or_index:
+            return parent_menu
+    raise ValueError(
+        f"Parent with {'position' if is_lookup_by_position else 'label'} '{lookup_label_or_index}' not found"
+    )
+
+
+def register_dashboard_menu(menu_or_function=None, parent_id=None):
+    if menu_or_function is None:
+        def decorator(decorated_function):
+            register_dashboard_menu(decorated_function, parent_id)
+            return decorated_function
+
+        return decorator
+
+    menu = menu_or_function() if callable(menu_or_function) else menu_or_function
+    if menu is None:
+        return
+
+    if isinstance(menu, dict):
+        menu = Menu.from_dict(menu)
+
+    if parent_id is None:
+        try:
+            _dashboard_navigation_menu[menu.identifier] += menu
+        except TypeError:
+            # `defaultdict` tried to instantiate a default Menu object but could not because
+            # 'label' positional argument is required
+            menu.auto_position_if_applicable(len(_dashboard_navigation_menu))
+            _dashboard_navigation_menu[menu.identifier] = menu
+    else:
+        # Placeholder instance will be used temporarily if child (menu) was registered ahead of
+        # its parent. It will be replaced (merged) into the actual parent once it's registered.
+        _dashboard_navigation_menu.setdefault(parent_id, Menu.placeholder(parent_id)).add_child(
+            menu).auto_position_if_applicable(len(_dashboard_navigation_menu))
```

Comment:
Can you add comments to the above code and explain why there are so many different types of object accepted as input to create a menu?